pkg/collector/edp/collector_test.go

```diff
@@ -0,0 +1,471 @@
+package edp
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"testing"
+
+	"github.com/google/uuid"
+	"github.com/onsi/gomega"
+	"github.com/prometheus/client_golang/prometheus/testutil"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap/zapcore"
+	"k8s.io/utils/ptr"
+
+	"github.com/kyma-project/kyma-metrics-collector/pkg/collector"
+	"github.com/kyma-project/kyma-metrics-collector/pkg/collector/edp/stubs"
+	"github.com/kyma-project/kyma-metrics-collector/pkg/logger"
+	"github.com/kyma-project/kyma-metrics-collector/pkg/resource"
+	"github.com/kyma-project/kyma-metrics-collector/pkg/runtime"
+	kmctesting "github.com/kyma-project/kyma-metrics-collector/pkg/testing"
+)
+
+const (
+	testDataStream = "dataStream"
+	retryCount     = 1
+	backendName    = "edp"
+)
+
+func TestCollector_CollectAndSend(t *testing.T) {
+	scannerID1 := resource.ScannerID("scanner1")
+	scannerID2 := resource.ScannerID("scanner2")
+
+	// In order to test the aggregation of EDP measurements, we are going to executes the tests with two scanners: scanner1 and scanner2.
+	// scanner1 will behave the same in all test cases: successful scan and successful conversion to EDP measurement.
+	// scanner2 will behave differently in each test case:
+	// case 1: scanner2 succeeds in scanning and conversion to EDP measurement succeeds.
+	// case 2: scanner2 fails in scanning, but previous scan exists. So, the previous scan will be used for conversion to EDP measurement.
+	// case 3: scanner2 fails in scanning and previous scan doesn't exist. So, nothing else we can do here and payload will be sent to EDP without scanner2's data.
+	// case 4: scanner2 succeeds in scanning, but conversion to EDP measurement fails. So, the previous scan will be used for conversion to EDP measurement.
+	// case 5: scanner2 succeeds in scanning, but conversion to EDP measurement fails and previous scan doesn't exist. So, nothing else we can do here and payload will be sent to EDP without scanner2's data.
+	// case 6: scanner2 succeeds in scanning, but conversion to EDP measurement fails and previous scan exists, but conversion of previous scan to EDP measurement also fails. So, nothing else we can do here and payload will be sent to EDP without scanner2's data.
```

Comment:
is this even a valid scenario?
if conversion fails it should not be stored in the "previous" map

Reply:
lets remove this test case and verify in a test that this scenario cannot happen

Reply:
I removed test case 6.
And I make sure now that the scanMap returned by `CollectAndSend` contains only scans which succeed in the 2 steps (scanning and converting).
And case 5 already ensure now that the returned scan map will not contain scan from scanner2